File: tests/test_http_server_comprehensive.py
#!/usr/bin/env python3
"""
自研HTTP服务器全面测试脚本
使用Python requests库进行测试
"""

import requests
import json
import time
import threading
from concurrent.futures import ThreadPoolExecutor, as_completed
from typing import Dict, List, Tuple
import sys

class TestStats:
    def __init__(self):
        self.total = 0
        self.passed = 0
        self.failed = 0
        self.lock = threading.Lock()
    
    def add_result(self, passed: bool):
        with self.lock:
            self.total += 1
            if passed:
                self.passed += 1
            else:
                self.failed += 1
    
    def print(self):
        print("\n" + "=" * 50)
        print("测试统计:")
        print(f"  总计: {self.total}")
        print(f"  通过: {self.passed}")
        print(f"  失败: {self.failed}")
        if self.total > 0:
            print(f"  成功率: {100.0 * self.passed / self.total:.2f}%")
        print("=" * 50 + "\n")

class HttpClient:
    def __init__(self, base_url: str, timeout: int = 30):
        self.base_url = base_url.rstrip('/')
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update({'Content-Type': 'application/json'})
    
    def get(self, path: str) -> Tuple[int, str]:
        """GET请求"""
        try:
            url = self.base_url + path
            response = self.session.get(url, timeout=self.timeout)
            return response.status_code, response.text
        except Exception as e:
            return 0, str(e)
    
    def post(self, path: str, data: dict) -> Tuple[int, str]:
        """POST请求"""
        try:
            url = self.base_url + path
            response = self.session.post(url, json=data, timeout=self.timeout)
            return response.status_code, response.text
        except Exception as e:
            return 0, str(e)

# 测试统计
stats = TestStats()

def test(name: str, func):
    """测试辅助函数"""
    print(f"[TEST] {name}...", end='', flush=True)
    try:
        result = func()
        if result:
            print(" ✓ PASSED")
            stats.add_result(True)
        else:
            print(" ✗ FAILED")
            stats.add_result(False)
    except Exception as e:
        print(f" ✗ FAILED (exception: {e})")
        stats.add_result(False)

# ============================================================================
# 测试用例
# ============================================================================

def test_health_endpoint(client: HttpClient) -> bool:
    """测试1: 健康检查端点"""
    status_code, response = client.get("/health")
    if status_code != 200:
        return False
    try:
        data = json.loads(response)
        return data.get("success") == True
    except:
        return False

def test_generate_basic(client: HttpClient) -> bool:
    """测试2: 生成端点（基本）"""
    status_code, response = client.post("/generate", {
        "prompt": "Hello",
        "max_tokens": 5
    })
    if status_code != 200:
        return False
    try:
        data = json.loads(response)
        return data.get("success") == True and "data" in data and "text" in data["data"]
    except:
        return False

def test_encode_endpoint(client: HttpClient) -> bool:
    """测试3: 编码端点"""
    status_code, response = client.post("/encode", {
        "text": "Hello world"
    })
    if status_code != 200:
        return False
    try:
        data = json.loads(response)
        return data.get("success") == True and "data" in data and "tokens" in data["data"]
    except:
        return False

def test_404_not_found(client: HttpClient) -> bool:
    """测试4: 404 Not Found"""
    status_code, _ = client.get("/nonexistent")
    return status_code == 404

def test_invalid_json(client: HttpClient) -> bool:
    """测试5: 无效JSON"""
    try:
        url = client.base_url + "/generate"
        response = client.session.post(url, data="{invalid json}", timeout=client.timeout)
        return response.status_code == 400
    except:
        return False

def test_empty_request_body(client: HttpClient) -> bool:
    """测试6: 空请求体"""
    try:
        url = client.base_url + "/generate"
        response = client.session.post(url, data="", timeout=client.timeout)
        return response.status_code == 400
    except:
        return False

def test_large_request_body(client: HttpClient) -> bool:
    """测试7: 大请求体"""
    large_prompt = "A" * 10000  # 10KB
    status_code, _ = client.post("/generate", {
        "prompt": large_prompt,
        "max_tokens": 5
    })
    # 可能被拒绝，但不应崩溃
    return status_code in [200, 400]

def test_concurrent_requests(client: HttpClient, num_threads: int = 10, requests_per_thread: int = 10) -> bool:
    """测试8: 并发请求"""
    success_count = 0
    fail_count = 0
    
    def worker(thread_id: int):
        nonlocal success_count, fail_count
        for i in range(requests_per_thread):
            status_code, response = client.post("/generate", {
                "prompt": f"Test {thread_id}-{i}",
                "max_tokens": 3
            })
            if status_code == 200:
                try:
                    data = json.loads(response)
                    if data.get("success") == True:
                        with stats.lock:
                            success_count += 1
                        return
                except:
                    pass
            with stats.lock:
                fail_count += 1
    
    threads = []
    for i in range(num_threads):
        t = threading.Thread(target=worker, args=(i,))
        threads.append(t)
        t.start()
    
    for t in threads:
        t.join()
    
    total = num_threads * requests_per_thread
    success_rate = success_count / total if total > 0 else 0
    
    print(f"  (并发: {num_threads}线程 × {requests_per_thread}请求 = {total}请求, "
          f"成功: {success_count}, 成功率: {success_rate*100:.1f}%)", end='', flush=True)
    
    return success_rate >= 0.95  # 95%成功率

def test_keep_alive(client: HttpClient) -> bool:
    """测试9: Keep-Alive连接"""
    for i in range(5):
        status_code, _ = client.get("/health")
        if status_code != 200:
            return False
        time.sleep(0.01)
    return True

def test_stress_test(client: HttpClient, num_requests: int = 100) -> bool:
    """测试10: 压力测试"""
    success_count = 0
    fail_count = 0
    
    start_time = time.time()
    
    def worker(request_id: int):
        nonlocal success_count, fail_count
        status_code, response = client.post("/generate", {
            "prompt": f"Stress test {request_id}",
            "max_tokens": 3
        })
        if status_code == 200:
            try:
                data = json.loads(response)
                if data.get("success") == True:
                    with stats.lock:
                        success_count += 1
                    return
            except:
                pass
        with stats.lock:
            fail_count += 1
    
    threads = []
    for i in range(num_requests):
        t = threading.Thread(target=worker, args=(i,))
        threads.append(t)
        t.start()
    
    for t in threads:
        t.join()
    
    end_time = time.time()
    duration = end_time - start_time
    total = num_requests
    success_rate = success_count / total if total > 0 else 0
    qps = total / duration if duration > 0 else 0
    
    print(f"  (请求数: {num_requests}, 成功: {success_count}, "
          f"耗时: {duration:.2f}s, QPS: {qps:.1f}, "
          f"成功率: {success_rate*100:.1f}%)", end='', flush=True)
    
    return success_rate >= 0.90  # 90%成功率

def test_benchmark_endpoint(client: HttpClient) -> bool:
    """测试11: Benchmark端点"""
    status_code, response = client.post("/benchmark", {
        "requests": 5,
        "concurrency": 2,
        "max_tokens": 10,
        "prompt": "Test"
    })
    if status_code != 200:
        return False
    try:
        data = json.loads(response)
        return data.get("success") == True
    except:
        return False

def test_stream_endpoint(client: HttpClient) -> bool:
    """测试12: 流式端点（如果支持）"""
    try:
        url = client.base_url + "/generate_stream"
        response = client.session.post(url, json={
            "prompt": "Hello",
            "max_tokens": 5,
            "stream": True
        }, stream=True, timeout=client.timeout)
        return response.status_code == 200
    except:
        return False

# ============================================================================
# 主函数
# ============================================================================

def main():
    server_url = "http://localhost:8080"
    
    if len(sys.argv) > 1:
        server_url = sys.argv[1]
    
    print("=" * 50)
    print("自研HTTP服务器全面测试")
    print("=" * 50)
    print(f"服务器地址: {server_url}")
    print("开始测试...")
    print()
    
    client = HttpClient(server_url)
    
    # 等待服务器启动
    print("等待服务器就绪...", end='', flush=True)
    for i in range(10):
        status_code, _ = client.get("/health")
        if status_code == 200:
            print(" ✓")
            break
        time.sleep(1)
        print(".", end='', flush=True)
    else:
        print(" ✗ 服务器未就绪")
        return 1
    print()
    
    # 基本功能测试
    print("\n[基本功能测试]")
    test("健康检查端点", lambda: test_health_endpoint(client))
    test("生成端点（基本）", lambda: test_generate_basic(client))
    test("编码端点", lambda: test_encode_endpoint(client))
    test("Benchmark端点", lambda: test_benchmark_endpoint(client))
    test("流式端点", lambda: test_stream_endpoint(client))
    
    # 错误处理测试
    print("\n[错误处理测试]")
    test("404 Not Found", lambda: test_404_not_found(client))
    test("无效JSON", lambda: test_invalid_json(client))
    test("空请求体", lambda: test_empty_request_body(client))
    
    # 边界条件测试
    print("\n[边界条件测试]")
    test("大请求体", lambda: test_large_request_body(client))
    
    # 并发测试
    print("\n[并发测试]")
    test("并发请求（10线程×10请求）", lambda: test_concurrent_requests(client, 10, 10))
    test("并发请求（20线程×5请求）", lambda: test_concurrent_requests(client, 20, 5))
    test("并发请求（50线程×2请求）", lambda: test_concurrent_requests(client, 50, 2))
    
    # 连接测试
    print("\n[连接测试]")
    test("Keep-Alive连接", lambda: test_keep_alive(client))
    
    # 压力测试
    print("\n[压力测试]")
    test("压力测试（100请求）", lambda: test_stress_test(client, 100))
    test("压力测试（200请求）", lambda: test_stress_test(client, 200))
    test("压力测试（500请求）", lambda: test_stress_test(client, 500))
    
    # 打印统计
    stats.print()
    
    return 1 if stats.failed > 0 else 0

if __name__ == "__main__":
    sys.exit(main())

File: tests/test_http_server_comprehensive.cpp
/**
 * @file test_http_server_comprehensive.cpp
 * @brief 自研HTTP服务器全面测试
 * @author cLLM Team
 * @date 2026-01-20
 */

#include <iostream>
#include <vector>
#include <thread>
#include <chrono>
#include <atomic>
#include <cassert>
#include <curl/curl.h>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

// 测试统计
struct TestStats {
    std::atomic<int> total{0};
    std::atomic<int> passed{0};
    std::atomic<int> failed{0};
    
    void reset() {
        total = 0;
        passed = 0;
        failed = 0;
    }
    
    void print() const {
        std::cout << "\n========================================" << std::endl;
        std::cout << "测试统计:" << std::endl;
        std::cout << "  总计: " << total.load() << std::endl;
        std::cout << "  通过: " << passed.load() << std::endl;
        std::cout << "  失败: " << failed.load() << std::endl;
        std::cout << "  成功率: " << (total.load() > 0 ? (100.0 * passed.load() / total.load()) : 0) << "%" << std::endl;
        std::cout << "========================================\n" << std::endl;
    }
};

static TestStats g_stats;

// CURL回调函数
static size_t WriteCallback(void* contents, size_t size, size_t nmemb, std::string* data) {
    size_t totalSize = size * nmemb;
    data->append((char*)contents, totalSize);
    return totalSize;
}

// HTTP请求辅助函数
class HttpClient {
public:
    HttpClient(const std::string& baseUrl) : baseUrl_(baseUrl) {
        curl_global_init(CURL_GLOBAL_DEFAULT);
    }
    
    ~HttpClient() {
        curl_global_cleanup();
    }
    
    bool get(const std::string& path, int& statusCode, std::string& response) {
        CURL* curl = curl_easy_init();
        if (!curl) return false;
        
        std::string url = baseUrl_ + path;
        std::string data;
        
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &data);
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 30L);
        
        CURLcode res = curl_easy_perform(curl);
        
        if (res == CURLE_OK) {
            curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &statusCode);
            response = data;
        }
        
        curl_easy_cleanup(curl);
        return res == CURLE_OK;
    }
    
    bool post(const std::string& path, const std::string& body, int& statusCode, std::string& response) {
        CURL* curl = curl_easy_init();
        if (!curl) return false;
        
        std::string url = baseUrl_ + path;
        std::string data;
        
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, body.c_str());
        curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, body.length());
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &data);
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 30L);
        
        struct curl_slist* headers = nullptr;
        headers = curl_slist_append(headers, "Content-Type: application/json");
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
        
        CURLcode res = curl_easy_perform(curl);
        
        if (res == CURLE_OK) {
            curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &statusCode);
            response = data;
        }
        
        curl_slist_free_all(headers);
        curl_easy_cleanup(curl);
        return res == CURLE_OK;
    }
    
private:
    std::string baseUrl_;
};

// 测试辅助宏
#define TEST(name, func) \
    do { \
        std::cout << "[TEST] " << name << "..." << std::flush; \
        g_stats.total++; \
        try { \
            if (func()) { \
                std::cout << " ✓ PASSED" << std::endl; \
                g_stats.passed++; \
            } else { \
                std::cout << " ✗ FAILED" << std::endl; \
                g_stats.failed++; \
            } \
        } catch (const std::exception& e) { \
            std::cout << " ✗ FAILED (exception: " << e.what() << ")" << std::endl; \
            g_stats.failed++; \
        } \
    } while(0)

// ============================================================================
// 测试用例
// ============================================================================

// 测试1: 健康检查端点
bool test_health_endpoint(HttpClient& client) {
    int statusCode;
    std::string response;
    
    if (!client.get("/health", statusCode, response)) {
        return false;
    }
    
    if (statusCode != 200) {
        return false;
    }
    
    try {
        json j = json::parse(response);
        return j.contains("success") && j["success"] == true;
    } catch (...) {
        return false;
    }
}

// 测试2: 生成端点（基本）
bool test_generate_basic(HttpClient& client) {
    int statusCode;
    std::string response;
    
    json request;
    request["prompt"] = "Hello";
    request["max_tokens"] = 5;
    
    if (!client.post("/generate", request.dump(), statusCode, response)) {
        return false;
    }
    
    if (statusCode != 200) {
        return false;
    }
    
    try {
        json j = json::parse(response);
        return j.contains("success") && j["success"] == true &&
               j.contains("data") && j["data"].contains("text");
    } catch (...) {
        return false;
    }
}

// 测试3: 编码端点
bool test_encode_endpoint(HttpClient& client) {
    int statusCode;
    std::string response;
    
    json request;
    request["text"] = "Hello world";
    
    if (!client.post("/encode", request.dump(), statusCode, response)) {
        return false;
    }
    
    if (statusCode != 200) {
        return false;
    }
    
    try {
        json j = json::parse(response);
        return j.contains("success") && j["success"] == true &&
               j.contains("data") && j["data"].contains("tokens");
    } catch (...) {
        return false;
    }
}

// 测试4: 404 Not Found
bool test_404_not_found(HttpClient& client) {
    int statusCode;
    std::string response;
    
    if (!client.get("/nonexistent", statusCode, response)) {
        return false;
    }
    
    return statusCode == 404;
}

// 测试5: 无效JSON
bool test_invalid_json(HttpClient& client) {
    int statusCode;
    std::string response;
    
    std::string invalidJson = "{invalid json}";
    if (!client.post("/generate", invalidJson, statusCode, response)) {
        return false;
    }
    
    return statusCode == 400;  // Bad Request
}

// 测试6: 并发请求
bool test_concurrent_requests(HttpClient& client, int numThreads = 10, int requestsPerThread = 10) {
    std::atomic<int> successCount{0};
    std::atomic<int> failCount{0};
    
    auto worker = [&](int threadId) {
        for (int i = 0; i < requestsPerThread; ++i) {
            int statusCode;
            std::string response;
            
            json request;
            request["prompt"] = "Test " + std::to_string(threadId) + "-" + std::to_string(i);
            request["max_tokens"] = 3;
            
            if (client.post("/generate", request.dump(), statusCode, response)) {
                if (statusCode == 200) {
                    try {
                        json j = json::parse(response);
                        if (j.contains("success") && j["success"] == true) {
                            successCount++;
                        } else {
                            failCount++;
                        }
                    } catch (...) {
                        failCount++;
                    }
                } else {
                    failCount++;
                }
            } else {
                failCount++;
            }
        }
    };
    
    std::vector<std::thread> threads;
    for (int i = 0; i < numThreads; ++i) {
        threads.emplace_back(worker, i);
    }
    
    for (auto& t : threads) {
        t.join();
    }
    
    int total = numThreads * requestsPerThread;
    int success = successCount.load();
    double successRate = (double)success / total;
    
    std::cout << "  (并发: " << numThreads << "线程 × " << requestsPerThread 
              << "请求 = " << total << "请求, 成功: " << success 
              << ", 成功率: " << (successRate * 100) << "%)" << std::flush;
    
    return successRate >= 0.95;  // 95%成功率
}

// 测试7: Keep-Alive连接
bool test_keep_alive(HttpClient& client) {
    // 发送多个请求，验证连接复用
    for (int i = 0; i < 5; ++i) {
        int statusCode;
        std::string response;
        
        if (!client.get("/health", statusCode, response)) {
            return false;
        }
        
        if (statusCode != 200) {
            return false;
        }
        
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }
    
    return true;
}

// 测试8: 大请求体
bool test_large_request_body(HttpClient& client) {
    int statusCode;
    std::string response;
    
    json request;
    std::string largePrompt(10000, 'A');  // 10KB prompt
    request["prompt"] = largePrompt;
    request["max_tokens"] = 5;
    
    if (!client.post("/generate", request.dump(), statusCode, response)) {
        return false;
    }
    
    return statusCode == 200 || statusCode == 400;  // 可能被拒绝，但不应崩溃
}

// 测试9: 空请求体
bool test_empty_request_body(HttpClient& client) {
    int statusCode;
    std::string response;
    
    if (!client.post("/generate", "", statusCode, response)) {
        return false;
    }
    
    return statusCode == 400;  // Bad Request
}

// 测试10: 压力测试
bool test_stress_test(HttpClient& client, int numRequests = 100) {
    std::atomic<int> successCount{0};
    std::atomic<int> failCount{0};
    
    auto start = std::chrono::steady_clock::now();
    
    for (int i = 0; i < numRequests; ++i) {
        int statusCode;
        std::string response;
        
        json request;
        request["prompt"] = "Stress test " + std::to_string(i);
        request["max_tokens"] = 3;
        
        if (client.post("/generate", request.dump(), statusCode, response)) {
            if (statusCode == 200) {
                successCount++;
            } else {
                failCount++;
            }
        } else {
            failCount++;
        }
    }
    
    auto end = std::chrono::steady_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();
    
    int success = successCount.load();
    double successRate = (double)success / numRequests;
    double qps = (double)numRequests * 1000 / duration;
    
    std::cout << "  (请求数: " << numRequests << ", 成功: " << success 
              << ", 耗时: " << duration << "ms, QPS: " << qps 
              << ", 成功率: " << (successRate * 100) << "%)" << std::flush;
    
    return successRate >= 0.90;  // 90%成功率
}

// ============================================================================
// 主函数
// ============================================================================

int main(int argc, char* argv[]) {
    std::string serverUrl = "http://localhost:8080";
    
    if (argc > 1) {
        serverUrl = argv[1];
    }
    
    std::cout << "========================================" << std::endl;
    std::cout << "自研HTTP服务器全面测试" << std::endl;
    std::cout << "========================================" << std::endl;
    std::cout << "服务器地址: " << serverUrl << std::endl;
    std::cout << "开始测试..." << std::endl;
    std::cout << "" << std::endl;
    
    HttpClient client(serverUrl);
    
    // 等待服务器启动
    std::cout << "等待服务器就绪..." << std::flush;
    for (int i = 0; i < 10; ++i) {
        int statusCode;
        std::string response;
        if (client.get("/health", statusCode, response) && statusCode == 200) {
            std::cout << " ✓" << std::endl;
            break;
        }
        std::this_thread::sleep_for(std::chrono::seconds(1));
        std::cout << "." << std::flush;
    }
    std::cout << std::endl;
    
    // 基本功能测试
    std::cout << "\n[基本功能测试]" << std::endl;
    TEST("健康检查端点", [&]() { return test_health_endpoint(client); });
    TEST("生成端点（基本）", [&]() { return test_generate_basic(client); });
    TEST("编码端点", [&]() { return test_encode_endpoint(client); });
    
    // 错误处理测试
    std::cout << "\n[错误处理测试]" << std::endl;
    TEST("404 Not Found", [&]() { return test_404_not_found(client); });
    TEST("无效JSON", [&]() { return test_invalid_json(client); });
    TEST("空请求体", [&]() { return test_empty_request_body(client); });
    
    // 边界条件测试
    std::cout << "\n[边界条件测试]" << std::endl;
    TEST("大请求体", [&]() { return test_large_request_body(client); });
    
    // 并发测试
    std::cout << "\n[并发测试]" << std::endl;
    TEST("并发请求（10线程×10请求）", [&]() { return test_concurrent_requests(client, 10, 10); });
    TEST("并发请求（20线程×5请求）", [&]() { return test_concurrent_requests(client, 20, 5); });
    
    // 连接测试
    std::cout << "\n[连接测试]" << std::endl;
    TEST("Keep-Alive连接", [&]() { return test_keep_alive(client); });
    
    // 压力测试
    std::cout << "\n[压力测试]" << std::endl;
    TEST("压力测试（100请求）", [&]() { return test_stress_test(client, 100); });
    TEST("压力测试（200请求）", [&]() { return test_stress_test(client, 200); });
    
    // 打印统计
    g_stats.print();
    
    return g_stats.failed.load() > 0 ? 1 : 0;
}
